curvlinops/ggn.py

```diff
@@ -1,12 +1,117 @@
+    ) -> Tuple[Tensor, ...]:
+        """Multiply the GGN on a vector in list format.
+
+        Args:
+            params: Parameters of the model.
+            X: Input to the DNN.
+            y: Ground truth.
+            *v: Vector to be multiplied with in tensor list format.
+
+        Returns:
+            Result of GGN multiplication in list format. Has the same shape as
+            ``v``, i.e. each tensor in the list has the shape of a parameter.
+        """
+        # Apply the Jacobian of f onto v: v → Jv
+        f_val, f_jvp = jvp(lambda *params_inner: f(*params_inner, X), params, v)
+
+        # Apply the criterion's Hessian onto Jv: Jv → HJv
+        c_grad_func = jacrev(lambda pred: c(pred, y))
+        _, c_hvp = jvp(c_grad_func, (f_val,), (f_jvp,))
+
+        # Apply the transposed Jacobian of f onto HJv: HJv → JᵀHJv
+        # NOTE This re-evaluates the net's forward pass. [Unverified] It should be op-
+        # timized away by common sub-expression elimination if you compile the function.
+        _, f_vjp_func = vjp(lambda *params_inner: f(*params_inner, X), *params)
+        return f_vjp_func(c_hvp)
+
+    return ggn_vector_product
+
+
+def make_batch_ggn_matrix_product(
+    model_func: Module, loss_func: Module, params: Tuple[Parameter, ...]
+) -> Callable[
+    [Union[Tensor, MutableMapping], Tensor, Tuple[Tensor, ...]], Tuple[Tensor, ...]
+]:
+    r"""Set up function that multiplies the mini-batch GGN onto a matrix in list format.
+
+    Args:
+        model_func: The neural network :math:`f_{\mathbf{\theta}}`.
+        loss_func: The loss function :math:`\ell`.
+        params: A tuple of parameters w.r.t. which the GGN is computed.
+            All parameters must be part of ``model_func.parameters()``.
+
+    Returns:
+        A function that takes inputs ``X``, ``y``, and a matrix ``M`` in list
+        format, and returns the mini-batch GGN applied to ``M`` in list format.
+    """
+    # Create functional versions of the model (f: *params, X -> prediction) and
+    # criterion function (c: prediction, y -> loss)
+    f, c = make_functional_model_and_loss(model_func, loss_func, params)
+
+    # Create the functional GGN-vector product
+    ggn_vp = make_ggn_vector_product(f, c)
+
+    def ggn_vector_product(
+        X: Union[Tensor, MutableMapping], y: Tensor, *v: Tuple[Tensor, ...]
+    ) -> Tuple[Tensor, ...]:
+        """Multiply the mini-batch GGN on a vector in list format.
+
+        Args:
+            X: Input to the DNN.
+            y: Ground truth.
+            *v: Vector to be multiplied with in tensor list format.
+
+        Returns:
+            Result of GGN multiplication in list format. Has the same shape as
+            ``v``, i.e. each tensor in the list has the shape of a parameter.
+        """
+        return ggn_vp(params, X, y, *v)
+
+    # Vectorize over vectors to multiply onto a matrix in list format
+    return vmap(
+        ggn_vector_product,
```

Comment:
Is it possible to remove the whole `ggn_vector_product` function and replace this line with `functools.partial(ggn_vp, params=params)`?